components/encryption/src/master_key/file.rs

```diff
@@ -4,27 +4,60 @@ use kvproto::encryptionpb::{EncryptedContent, EncryptionMethod};
 
 use super::metadata::*;
 use super::Backend;
-use crate::crypter::*;
-use crate::{AesCtrCrypter, Error, Iv, Result};
+use crate::crypter::{self, AesCtrCrypter, Iv};
+use crate::{Error, Result};
+
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
 
 pub struct FileBackend {
     method: EncryptionMethod,
     key: Vec<u8>,
 }
 
 impl FileBackend {
-    pub fn new(method: EncryptionMethod, key: Vec<u8>) -> Result<FileBackend> {
-        if key.len() != get_method_key_length(method) {
-            return Err(Error::Other(
-                format!(
-                    "encryption method and key length mismatch, expect {} get {}",
-                    get_method_key_length(method),
-                    key.len()
-                )
-                .into(),
-            ));
-        }
-        Ok(FileBackend { key, method })
+    pub fn new<P: AsRef<Path>>(method: EncryptionMethod, path: P) -> Result<FileBackend> {
+        let key = match method {
+            EncryptionMethod::Unknown => return Err(Error::UnknownEncryption),
+            EncryptionMethod::Plaintext => vec![],
+            _ => {
+                let key_len = crypter::get_method_key_length(method);
+                let mut file = File::open(path)?;
+                // Check file size to avoid reading a gigantic file accidentally.
+                let file_len = file.metadata()?.len() as usize;
+                if file_len != key_len * 2 + 1 {
+                    return Err(Error::Other(
+                        format!(
+                            "mismatch master key file size, expected {}, actual {}.",
+                            key_len * 2 + 1,
+                            file_len,
+                        )
+                        .into(),
+                    ));
+                }
+                let mut content = vec![];
+                let read_len = file.read_to_end(&mut content)?;
+                if read_len != file_len {
```

Comment:
Is it necessary? If length can't match, an error should be reported in `read_to_end`.

Reply:
I briefly check `read_to_end` implementation and doesn't see it checks against file size. Doesn't hurt to be a bit paranoid.